server/src/main/java/keywhiz/commands/BackfillOwnershipCommand.java

```diff
@@ -0,0 +1,165 @@
+        .type(String.class)
+        .setDefault("PT0S")
+        .help("Delay between batches in ISO 8601 duration format e.g. PT0.1s");
+  }
+
+  @Override protected void run(Bootstrap<KeywhizConfig> bootstrap, Namespace namespace,
+      KeywhizConfig config) throws Exception {
+
+    ManagedDataSource dataSource = config.getDataSourceFactory().build(new MetricRegistry(), "backfill-ownership-datasource");
+    Injector injector = InjectorFactory.createInjector(
+        config,
+        Environments.fromBootstrap(bootstrap),
+        dataSource);
+
+    DSLContext jooq = injector.getInstance(Key.get(DSLContext.class, Readonly.class));
+    int batchSize = getBatchSize(namespace);
+    Duration delay = getDelay(namespace);
+
+    new Backfill(jooq, batchSize, delay).run();
+  }
+
+  private static int getBatchSize(Namespace namespace) {
+    Integer batchSize = namespace.getInt(Args.BATCH_SIZE);
+    return batchSize == null
+        ? 1
+        : batchSize;
+  }
+
+  private static Duration getDelay(Namespace namespace) {
+    String delayToken = namespace.getString(Args.DELAY);
+    return delayToken == null
+        ? Duration.ZERO
+        : Duration.parse(delayToken);
+  }
+
+  private static class Backfill implements Runnable {
+    private final DSLContext jooq;
+    private final int batchSize;
+    private final Duration delay;
+
+    public Backfill(
+        DSLContext jooq,
+        int batchSize,
+        Duration delay) {
+      this.jooq = jooq;
+      this.batchSize = batchSize;
+      this.delay = delay;
+    }
+
+    public void run() {
+      Cursor<Record1<Long>> cursor = allSecretsWithoutOwners();
+      try {
+        while (cursor.hasNext()) {
+          processBatch(cursor);
+          System.out.print(".");
+          sleepQuietly();
+        }
+      } finally {
+        if (!cursor.isClosed()) {
+          cursor.close();
+        }
+      }
+      System.out.println();
+    }
+
+    private void processBatch(Cursor<Record1<Long>> cursor) {
+      for (int i = 0; i < batchSize; i++) {
+        if (cursor.hasNext()) {
+          findAndUpdateOwner(cursor.fetchNext());
+        } else {
+          return;
+        }
+      }
+    }
+
+    private void sleepQuietly() {
+      try {
+        TimeUnit.MILLISECONDS.sleep(delay.toMillis());
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
+      }
+    }
+
+    private void findAndUpdateOwner(Record1<Long> record) {
+      long secretId = record.value1();
+      Optional<Long> maybeSecretOwner = getSecretOwner(secretId);
+      if (maybeSecretOwner.isPresent()) {
+        updateOwner(secretId, maybeSecretOwner.get());
+      }
+    }
+
+    private Cursor<Record1<Long>> allSecretsWithoutOwners() {
+      return jooq
+          .select(SECRETS.ID)
+          .from(SECRETS)
+          .where(SECRETS.OWNER.isNull())
+          .fetchSize(batchSize)
+          .fetchLazy();
+    }
```

Comment:
Originally I had a "forSharing()" step in here, with the idea to explicitly tell MySQL not to hold any locks. However, during testing the database yelled at me saying this was unsupported syntax. Odd. In any case it's removed, but the intent of the cursor is that we can scroll through results without having to hold a ton of locks. I need to do a bit more reading on the MySQL side to see if there's another hint we can provide.